sat/apiclient/pcs.py

```diff
@@ -0,0 +1,195 @@
+                                           f'e={err_info[0]} and err_msg="{err_info[1]}"'
+                                           for err_info, xnames in xnames_by_err.items()])
+
+            return f'{self.message}\n{xname_err_summary}'
+
+
+class PCSClient(APIGatewayClient):
+    """Client for the Power Control Service."""
+    base_resource_path = 'power-control/v1/'
+
+    def set_xnames_power_state(self, xnames, power_state, force=False, recursive=False, prereq=False):
+        """Set the power state of the given xnames.
+
+        Args:
+            xnames (list): the xnames (str) to perform the power operation
+                against.
+            power_state (str): the desired power state. Either "on" or "off".
+            force (bool): if True, disable checks and force the power operation.
+            recursive (bool): if True, power on component and its descendants.
+            prereq (bool): if True, power on component and its ancestors.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: if the given `power_state` is not one of 'on' or 'off'
+            PCSError: if the attempt to power on/off the given xnames with PCS
+                fails. This exception contains more specific information about
+                the failure, which will be included in its __str__.
+        """
+        allowed_states = {'on', 'off', 'soft-off', 'soft-restart', 'hard-restart', 'init', 'force-off'}
+        power_state = power_state.lower()
+        if power_state not in allowed_states:
+            allowed_states_str = ", ".join("\"" + state + "\"" for state in allowed_states)
+            raise ValueError(f'Invalid power state {power_state} given. Must be {allowed_states_str}')
+
+        if force:
+            if power_state in {'off', 'soft-off'}:
+                power_state = 'force-off'
+            elif power_state == 'soft-restart':
+                power_state = 'hard-restart'
+
+        target_xnames = set()
+        if recursive:
+            hsm_client = HSMClient(self.session)
+
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for xname in xnames
+                    for component in hsm_client.get_node_components(ancestor=xname)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not retrieve descendent components for xnames: {err}') from err
+        if prereq:
+            hsm_client = HSMClient(self.session)
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for component in hsm_client.get_all_components()
+                    if any(xname.startswith(component['XName']) for xname in xnames)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not query ancestor components for xnames: {err}') from err
```

Comment:
This `prereq` implementation also makes me nervous because it's quite different from the way it's done in [add_parents in the cray CLI](https://github.com/Cray-HPE/craycli/blob/082288ed1813f89acb2007f8f7d42cf68ea0aa29/cray/modules/power/cli.py#L128).

I also don't see anything that actually relies on this functionality, so perhaps we can just remove it.

Reply:
addressed